src/python/zquantum/qaoa/recursive_qaoa.py

```diff
@@ -0,0 +1,394 @@
+    qubit_itll_be_replaced_with: int = term_with_largest_expval[0][0]
+
+    new_indice = old_indice
+
+    if old_indice > qubit_to_get_rid_of:
+        # map qubit to the qubit 1 below it
+        new_indice = old_indice - 1
+    elif old_indice == qubit_to_get_rid_of:
+        # map qubit onto the qubit it's being replaced with
+        new_indice = qubit_itll_be_replaced_with
+
+    return new_indice
+
+
+def _create_reduced_hamiltonian(
+    hamiltonian: IsingOperator,
+    term_with_largest_expval: IsingOperator,
+    largest_expval: float,
+) -> IsingOperator:
+    """Reduce the cost hamiltonian by substituting one qubit of the term with largest expectation
+    value with the other qubit of the term. See equation (15) of the original paper.
+
+    Args:
+        hamiltonian: hamiltonian to be reduced
+        term_with_largest_expval: term with largest expectation value
+        largest_expval: the expectation value of `term_with_largest_expval`
+
+    Returns:
+        Reduced hamiltonian.
+    """
+    assert len(term_with_largest_expval.terms.keys()) == 1
+
+    reduced_hamiltonian = IsingOperator()
+
+    qubit_to_get_rid_of: int = [*term_with_largest_expval.terms][0][1][0]
+
+    for (term, coefficient) in hamiltonian.terms.items():
+        # term is tuple representing one term of IsingOperator, example ((2, 'Z'), (3, 'Z'))
+        if term not in term_with_largest_expval.terms:
+            # If term is not the term_with_largest_expval
+            new_term: Tuple = ()
+            for qubit in term:
+                # qubit is a component of qubit operator on 1 qubit ex. (2, 'Z')
+                qubit_indice: int = qubit[0]
+
+                # Map the new cost hamiltonian onto reduced qubits
+                new_qubit_indice = _get_new_qubit_indice(
+                    qubit_indice, term_with_largest_expval
+                )
+                new_qubit = (new_qubit_indice, "Z")
+                new_term += (new_qubit,)
+
+                if qubit_indice == qubit_to_get_rid_of:
+                    coefficient *= np.sign(largest_expval)
+
+            reduced_hamiltonian += IsingOperator(new_term, coefficient)
+
+    return reduced_hamiltonian
+
+
+def _map_reduced_solutions_to_original_solutions(
+    reduced_solutions: List[Tuple[int]], qubit_map: List[List[int]]
+):
+    """Maps the answer of the reduced Hamiltonian back to the original number of qubits.
+
+    Args:
+        reduced_solutions: list of solutions, each solution is a tuple of ints.
+        qubit_map: list that maps original qubits to new qubits, see docstring of RecursiveQAOA
+
+    Returns:
+        list of solutions, each solution is a tuple of ints.
+    """
+
+    original_solutions: List[Tuple[int, ...]] = []
+
+    for reduced_solution in reduced_solutions:
+        original_solution: List[int] = []
+        for qubit in qubit_map:
+            this_answer = reduced_solution[np.abs(qubit[0])]
```

Comment:
Why is np.abs needed here?

Reply:
Good observation, it's not needed anymore. I'll remove it